pkg/controller/podprobemarker/ppm_event_handler.go

```diff
@@ -0,0 +1,248 @@
+}
+
+func (p *enqueueRequestForPodProbeMarker) Update(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	p.queue(q, evt.ObjectNew)
+}
+
+func (p *enqueueRequestForPodProbeMarker) queue(q workqueue.RateLimitingInterface, obj runtime.Object) {
+	ppm, ok := obj.(*appsalphav1.PodProbeMarker)
+	if !ok {
+		return
+	}
+	q.Add(reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: ppm.Namespace,
+			Name:      fmt.Sprintf("%s#%s", ReconPodProbeMarker, ppm.Name),
+		},
+	})
+}
+
+var _ handler.EventHandler = &enqueueRequestForNodePodProbe{}
+
+type enqueueRequestForNodePodProbe struct{}
+
+func (p *enqueueRequestForNodePodProbe) Create(evt event.CreateEvent, q workqueue.RateLimitingInterface) {
+	obj, ok := evt.Object.(*appsalphav1.NodePodProbe)
+	if !ok {
+		return
+	}
+	p.queue(q, obj)
+}
+
+func (p *enqueueRequestForNodePodProbe) Delete(evt event.DeleteEvent, q workqueue.RateLimitingInterface) {
+}
+
+func (p *enqueueRequestForNodePodProbe) Generic(evt event.GenericEvent, q workqueue.RateLimitingInterface) {
+}
+
+func (p *enqueueRequestForNodePodProbe) Update(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	new, ok := evt.ObjectNew.(*appsalphav1.NodePodProbe)
+	if !ok {
+		return
+	}
+	old, ok := evt.ObjectOld.(*appsalphav1.NodePodProbe)
+	if !ok {
+		return
+	}
+	if reflect.DeepEqual(new.Status, old.Status) {
+		return
+	}
+	p.queue(q, new)
+}
+
+func (p *enqueueRequestForNodePodProbe) queue(q workqueue.RateLimitingInterface, npp *appsalphav1.NodePodProbe) {
+	q.Add(reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: npp.Namespace,
+			Name:      fmt.Sprintf("%s#%s", ReconNodePodProbe, npp.Name),
+		},
+	})
+}
+
+var _ handler.EventHandler = &enqueueRequestForPod{}
+
+type enqueueRequestForPod struct {
+	reader client.Reader
+}
+
+func (p *enqueueRequestForPod) Create(evt event.CreateEvent, q workqueue.RateLimitingInterface) {
+}
+
+func (p *enqueueRequestForPod) Delete(evt event.DeleteEvent, q workqueue.RateLimitingInterface) {
+}
+
+func (p *enqueueRequestForPod) Generic(evt event.GenericEvent, q workqueue.RateLimitingInterface) {
+}
+
+func (p *enqueueRequestForPod) Update(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	new, ok := evt.ObjectNew.(*corev1.Pod)
+	if !ok {
+		return
+	}
+	old, ok := evt.ObjectOld.(*corev1.Pod)
+	if !ok {
+		return
+	}
+	// remove pod probe from nodePodProbe.spec
+	if old.DeletionTimestamp.IsZero() && !new.DeletionTimestamp.IsZero() {
+		if new.Spec.NodeName == "" {
+			return
+		}
+		npp := &appsalphav1.NodePodProbe{}
+		if err := p.reader.Get(context.TODO(), client.ObjectKey{Name: new.Spec.NodeName}, npp); err != nil {
+			return
+		}
+		q.Add(reconcile.Request{NamespacedName: types.NamespacedName{Name: fmt.Sprintf("%s#%s", ReconNodePodProbe, new.Spec.NodeName)}})
+		return
+	}
+
+	// add pod probe to nodePodProbe.spec
+	if old.Spec.NodeName == "" && new.Spec.NodeName != "" {
```

Comment:
maybe we can wait until all init containers in the pod have started successfully(Initialized condition)